### PR TITLE
Fix content type detection with leading whitespace

### PR DESCRIPTION
**Pull request by jasontedor**
Today content type detection on an input stream works by peeking up to twenty bytes into the stream. If the stream is headed by more whitespace than twenty bytes, we might fail to detect the content type. We should be ignoring this whitespace before attempting to detect the content type. This commit does that by ignoring all leading whitespace in an input stream before attempting to guess the content type.

Relates #32357

### CODE REVIEW

**Comment by andrershov** [comment]
@jasontedor what is the reason to constantly double readLimit in `si.mark()` call? Why not just to call 
`si.mark(Integer.MAX_VALUE-8)` and skip all whitespaces during the first pass?  I think all sane implementations allocate reading buffer lazily, increasing its capacity when data does not fit.

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java [hunk 1]
```diff
@@ -227,23 +229,49 @@ public static XContent xContent(byte[] data, int offset, int length) {
...
-        try {
-            final byte[] firstBytes = new byte[GUESS_HEADER_LENGTH];
-            int read = 0;
-            while (read < GUESS_HEADER_LENGTH) {
-                final int r = si.read(firstBytes, read, GUESS_HEADER_LENGTH - read);
-                if (r == -1) {
-                    break;
+        int iteration = 1;
+        while (true) {
+            si.mark(iteration * GUESS_HEADER_LENGTH);
```
Comment (bleskes) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java]: This is where is a bit uncomfortable with requiring the underlying stream to have an unbound support for mark. If I read things correctly we end up with a stream that just wraps a byte reference and we therefore don't care. If you're comfortable relying on that behavior, then I'm good but then I also think we don't need to allocate a growing size of `firstBytes` but rather read byte by byte until we find the first non-white space.

**Comment by jasontedor** [comment]
@andrershov It was to deal with possibility of insane implementations but it appears that we should always have a sane implementation here.

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java (hunk 1 above)
Comment (jasontedor) [reply libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java]: I am comfortable with relying on that behavior. I pushed 86227baa5a7c80eec1cd3328b1c75d0d48c54402.

**Review by bleskes — APPROVED** [review]
LGTM

**Pull request merged**